flax/nn/base.py

```diff
@@ -931,29 +940,62 @@ def store(self, value):
     Returns:
       The previous value stored in the collection or None.
     """
-    _top_frame('store')
+    frame = _top_frame('store')
     if not self._mutable:
       raise ValueError('Collection is not mutable. Use the `mutate` method to'
                        'create a mutable copy.')
-    assert len(_module_stack) > self._anchor
-
     # the root of a Collection is the first module scope that gets created
     # inside the mutate scope of the Collection. By allowing only one unique
     # root scope we guarantee that state is not accidentally shared
     # between different models. When a user specifies an explicit name we can
     # distinguish models and a collection can have multiple roots.
-    root = _module_stack[self._anchor]
+    if frame is self._anchor:
+      # Example:
+      # with nn.Collection.mutate() as coll:
+      #   coll.store(1)
+      raise ValueError('State should be stored from within a module.'
+                       ' Consider using the value directly instead of'
+                       ' storing it in a Collection.')
+    if not frame.is_child_of(self._anchor):
+      # edge case where the Collection cannot capture the scope of a shared Module
+      # See test_collection_store_fails_if_out_of_scope in nn_test.py
+      raise ValueError('Trying to capture state outside the scope of this Collection.'
+                       ' Most likely due to passing around a shared Module.')
+    root = self._find_root(frame)
     if self._root is None:
       self._root = root
-    elif self._root is not root:
+    elif self._root != root:
       if self._root.name is None or root.name is None:
+        # Example:
+        # with nn.Collection() as coll:
+        #   StatefulModule.call(params, coll)
+        #   StatefulModule.call(params2, coll)
         raise ValueError('When multiple top-level module calls use a Collection'
```

Comment:
Can you explain more about why this is a problem?  Is it because you want to make it explicit whether they share state or not? I spent a few minutes thinking about it and couldn't figure it out.